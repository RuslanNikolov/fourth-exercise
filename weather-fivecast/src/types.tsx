export interface DailyWeatherForecast {
    dayForecast: IDayForecast
    hourForecasts: IHourForecast[],
}

export interface IDayForecast {
    date:string,
    averageTemp: number,
    averageFeelsLike: number,
    averageHumidity: number,
    minTemp: number,
    maxTemp: number,
    averageCloudiness: number
    averageRainProbability: number
}

export interface IWeatherResponseData
 {
    city: {
        name: string,
        country: string,
        sunrise: number, // time of sunrise in UNIX
        sunset: number // time of sunrise in UNIX
        [otherKey: string]: any
    },
    list: IHourForecast[],
    [otherKey:string]: any
}

export interface IHourForecast {
    dt_txt: string // time of data forecasted, ISO, UTC 
    main: {
        temp: number // Temperature
        feels_like: number, // This temperature parameter accounts for the human perception of weather.
        temp_min: number // Minimum temperature at the moment of calculation.
        temp_max: number, // Maximum temperature at the moment of calculation. 
        pressure: number, // Atmospheric pressure on the sea level by default,
        grnd_level: number, // Atmospheric pressure on the ground level,
        humidity: number, // Humidity %
        [otherKey: string]: any
    },
    weather: { // array which only has 1 element- a weather description object
        main: string // Group of weather parameters (Rain, Snow, Extreme etc.),
        description?: string // Additional weather description
        [otherKey: string]: any
    }[],
    wind: {
        speed: number // Wind speed. Unit Default: meter/sec, Metric: meter/sec,
        deg: number // Wind direction, degrees (meteorological)
        [otherKey: string]: any
    },
    clouds: {
        all: number
    },
    pop: number // Probability of raininng (number from 0 to 1)
    [otherKey: string]: any
}