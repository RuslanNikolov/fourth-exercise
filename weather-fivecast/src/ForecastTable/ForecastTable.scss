.table-head-row > th {
    font-weight: bold;
}

.date-cell {
    font-size: 1.2rem;
}

