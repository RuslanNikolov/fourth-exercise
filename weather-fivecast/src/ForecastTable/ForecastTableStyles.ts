// in some instances (not on this project) its better to use a CSS-in-JS method,
// it gives you some flexibility like passing props to the style, etc
export const tableHeadRow = {
    fontWeight: 'bold'
}