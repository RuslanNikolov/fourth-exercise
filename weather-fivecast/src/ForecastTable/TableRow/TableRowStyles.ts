export const dateCell = {
    fontSize: '1.2rem'
}