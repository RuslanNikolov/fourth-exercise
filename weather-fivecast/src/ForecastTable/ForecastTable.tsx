import React from 'react';
import { TableContainer, Table, TableBody, TableHead, TableRow as MUITableRow, TableCell, Typography,Paper } from '@mui/material';
import { DailyWeatherForecast } from '../types'
import './ForecastTableStyles.ts';
import TableRow from './TableRow/TableRow';
import * as styles from './ForecastTableStyles';

interface IForecastTableProps {
  currentCity: string,
  dailyForecasts: DailyWeatherForecast[]
}

const ForecastTable = (props: IForecastTableProps) => {

  return (
    <TableContainer component={Paper}>
      <Table aria-label="collapsible table" /* better red by screen readers */ >
        <TableHead >
          <MUITableRow sx={styles.tableHeadRow}>
            <TableCell />
            <TableCell>
              <Typography style={{ textAlign: "center", color: "green" }} variant="h3" >&#9774;</Typography>
            </TableCell>
            <TableCell>Temperature</TableCell>
            <TableCell>Feels like</TableCell>
            <TableCell>Minimium Temp</TableCell>
            <TableCell>Maximum Temp</TableCell>
            <TableCell>Humidity</TableCell>
            <TableCell>Cloudiness</TableCell>
            <TableCell>Rain probability</TableCell>
          </MUITableRow>
        </TableHead>

        <TableBody>
          {props.dailyForecasts.map(
            (dailyForecast: DailyWeatherForecast) => (
              <TableRow
                key={dailyForecast.hourForecasts[0].dt_txt} // an unique key on the map iteratior is a fundamental optimising tool
                dailyForecast={dailyForecast}
              />
            ))}
        </TableBody>
      </Table>
    </TableContainer >
  );
}

export default ForecastTable;
