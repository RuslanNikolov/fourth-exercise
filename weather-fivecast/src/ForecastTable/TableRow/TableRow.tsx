import React from 'react';
// we cannot do something like import * as MUI from @mui.. because that will import too much stuff
import { TableRow as MUITableRow, TableCell, Typography, IconButton, Collapse, Box, Table, TableHead, TableBody } from '@mui/material';
import { KeyboardArrowUp, KeyboardArrowDown } from '@mui/icons-material';
import { DailyWeatherForecast, IHourForecast } from '../../types';
import * as styles from './TableRowStyles';

interface ITableRowProps {
  dailyForecast: DailyWeatherForecast
}

const TableRow = (props: ITableRowProps) => {
  const [open, setOpen] = React.useState(false);

  return (
    <React.Fragment>
      <MUITableRow sx={{ '& > *': { borderBottom: 'unset' } }}>
        <TableCell sx={{ display: 'flex', flexDirection: 'column', alignItems: 'center', justifyContent: 'center' }}>
          <Typography variant="caption">BY HOURS</Typography>
          <IconButton style={styles.dateCell} aria-label="expand row" size="small" onClick={() => setOpen(!open)}>
            {open ? <KeyboardArrowUp /> : <KeyboardArrowDown />}
          </IconButton>
        </TableCell>
        <TableCell className="date-cell" component="th" scope="row">
          <strong>{props.dailyForecast.dayForecast.date}</strong>
        </TableCell>
        <TableCell>{props.dailyForecast.dayForecast.averageTemp}째</TableCell>
        <TableCell>{props.dailyForecast.dayForecast.averageFeelsLike}째</TableCell>
        <TableCell>{props.dailyForecast.dayForecast.minTemp}째</TableCell>
        <TableCell>{props.dailyForecast.dayForecast.maxTemp}째</TableCell>
        <TableCell>{props.dailyForecast.dayForecast.averageHumidity}</TableCell>
        <TableCell>{props.dailyForecast.dayForecast.averageCloudiness}</TableCell>
        <TableCell>{props.dailyForecast.dayForecast.averageRainProbability * 100}%</TableCell>
      </MUITableRow>

      <MUITableRow>
        <TableCell style={{ paddingBottom: 0, paddingTop: 0 }} colSpan={12}>
          <Collapse in={open} timeout="auto" unmountOnExit>
            <Box style={{ margin: 1 }}>
              <Typography variant="h6" gutterBottom component="div">
                Hourly details
              </Typography>
              <Table size="small" aria-label="purchases">
                <TableHead>
                  <MUITableRow>
                    <TableCell />
                    <TableCell>Temperature</TableCell>
                    <TableCell>Feels like</TableCell>
                    <TableCell>Atmospheric pressure</TableCell>
                    <TableCell>Humidity</TableCell>
                    <TableCell>Weather</TableCell>
                    <TableCell>Windspeed</TableCell>
                  </MUITableRow>
                </TableHead>
                <TableBody>
                  {props.dailyForecast.hourForecasts.map((hourForecast: IHourForecast) => (
                    <MUITableRow key={hourForecast.dt_txt}>
                      <TableCell component="th" scope="row">
                        {hourForecast.dt_txt.split(' ')[1]}
                      </TableCell>
                      <TableCell>{hourForecast.main.temp}</TableCell>
                      <TableCell>{hourForecast.main.feels_like}</TableCell>
                      <TableCell>{hourForecast.main.pressure}</TableCell>
                      <TableCell>{hourForecast.main.humidity}</TableCell>
                      <TableCell>{hourForecast.weather[0].main}</TableCell>
                      <TableCell>{hourForecast.wind.speed}</TableCell>
                    </MUITableRow>
                  ))}
                </TableBody>
              </Table>
            </Box>
          </Collapse>
        </TableCell>
      </MUITableRow>
    </React.Fragment>
  );
}

export default TableRow;


