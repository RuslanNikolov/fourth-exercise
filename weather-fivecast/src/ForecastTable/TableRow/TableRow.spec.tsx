import React from 'react';
import enzyme from 'enzyme';
import TableRow from './TableRow';
import { DailyWeatherForecast, IHourForecast } from '../../types'
import toJson from 'enzyme-to-json';
import {dailyForecastsMock} from '../ForecastTable.spec';

it('renders correctly', () => {
    const component = enzyme.shallow(<TableRow dailyForecast={dailyForecastsMock[0]} />);
    expect(toJson(component)).toMatchSnapshot();
});