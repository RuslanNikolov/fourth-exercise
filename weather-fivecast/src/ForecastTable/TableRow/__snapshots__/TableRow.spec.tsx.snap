// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<ForwardRef(TableContainer)
  component={
    Object {
      "$$typeof": Symbol(react.forward_ref),
      "propTypes": Object {
        "children": [Function],
        "className": [Function],
        "classes": [Function],
        "component": [Function],
        "elevation": [Function],
        "square": [Function],
        "sx": [Function],
        "variant": [Function],
      },
      "render": [Function],
    }
  }
>
  <ForwardRef(Table)
    aria-label="collapsible table"
  >
    <ForwardRef(TableHead)>
      <ForwardRef(TableRow)
        sx={
          Object {
            "fontWeight": "bold",
          }
        }
      >
        <ForwardRef(TableCell) />
        <ForwardRef(TableCell)>
          <ForwardRef(Typography)
            style={
              Object {
                "color": "green",
                "textAlign": "center",
              }
            }
            variant="h3"
          >
            ☮
          </ForwardRef(Typography)>
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)>
          Temperature
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)>
          Feels like
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)>
          Minimium Temp
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)>
          Maximum Temp
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)>
          Humidity
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)>
          Cloudiness
        </ForwardRef(TableCell)>
        <ForwardRef(TableCell)>
          Rain probability
        </ForwardRef(TableCell)>
      </ForwardRef(TableRow)>
    </ForwardRef(TableHead)>
    <ForwardRef(TableBody)>
      <TableRow
        dailyForecast={
          Object {
            "dayForecast": Object {
              "averageCloudiness": 34,
              "averageFeelsLike": 12,
              "averageHumidity": 84,
              "averageRainProbability": 0,
              "averageTemp": 12,
              "date": "2021-10-26",
              "maxTemp": 13,
              "minTemp": 12,
            },
            "hourForecasts": Array [
              Object {
                "clouds": Object {
                  "all": 56,
                },
                "dt_txt": "2021-10-26 12:00:00",
                "main": Object {
                  "feels_like": 11.93,
                  "grnd_level": 1017,
                  "humidity": 87,
                  "pressure": 1018,
                  "sea_level": 1018,
                  "temp": 12.37,
                  "temp_kf": -0.55,
                  "temp_max": 12.92,
                  "temp_min": 12.37,
                },
                "pop": 0,
                "visibility": 10000,
                "weather": Array [
                  Object {
                    "main": "Clouds",
                  },
                ],
                "wind": Object {
                  "deg": 2,
                  "speed": 2.09,
                },
              },
              Object {
                "clouds": Object {
                  "all": 12,
                },
                "dt_txt": "2021-10-27 12:00:00",
                "main": Object {
                  "feels_like": 11.98,
                  "grnd_level": 1018,
                  "humidity": 80,
                  "pressure": 1018,
                  "sea_level": 1018,
                  "temp": 12.38,
                  "temp_kf": -0.58,
                  "temp_max": 12.92,
                  "temp_min": 12.37,
                },
                "pop": 0.4,
                "visibility": 10000,
                "weather": Array [
                  Object {
                    "main": "Rain",
                  },
                ],
                "wind": Object {
                  "deg": 1,
                  "speed": 2.12,
                },
              },
            ],
          }
        }
        key="2021-10-26 12:00:00"
      />
    </ForwardRef(TableBody)>
  </ForwardRef(Table)>
</ForwardRef(TableContainer)>
`;

exports[`renders correctly 2`] = `
<Fragment>
  <ForwardRef(TableRow)
    sx={
      Object {
        "& > *": Object {
          "borderBottom": "unset",
        },
      }
    }
  >
    <ForwardRef(TableCell)
      sx={
        Object {
          "alignItems": "center",
          "display": "flex",
          "flexDirection": "column",
          "justifyContent": "center",
        }
      }
    >
      <ForwardRef(Typography)
        variant="caption"
      >
        BY HOURS
      </ForwardRef(Typography)>
      <ForwardRef(IconButton)
        aria-label="expand row"
        onClick={[Function]}
        size="small"
        style={
          Object {
            "fontSize": "1.2rem",
          }
        }
      >
        <Memo(ForwardRef(KeyboardArrowDownIcon)) />
      </ForwardRef(IconButton)>
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)
      className="date-cell"
      component="th"
      scope="row"
    >
      <strong>
        2021-10-26
      </strong>
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)>
      12
      °
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)>
      12
      °
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)>
      12
      °
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)>
      13
      °
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)>
      84
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)>
      34
    </ForwardRef(TableCell)>
    <ForwardRef(TableCell)>
      0
      %
    </ForwardRef(TableCell)>
  </ForwardRef(TableRow)>
  <ForwardRef(TableRow)>
    <ForwardRef(TableCell)
      colSpan={12}
      style={
        Object {
          "paddingBottom": 0,
          "paddingTop": 0,
        }
      }
    >
      <ForwardRef(Collapse)
        in={false}
        timeout="auto"
        unmountOnExit={true}
      >
        <ForwardRef(Box)
          style={
            Object {
              "margin": 1,
            }
          }
        >
          <ForwardRef(Typography)
            component="div"
            gutterBottom={true}
            variant="h6"
          >
            Hourly details
          </ForwardRef(Typography)>
          <ForwardRef(Table)
            aria-label="purchases"
            size="small"
          >
            <ForwardRef(TableHead)>
              <ForwardRef(TableRow)>
                <ForwardRef(TableCell) />
                <ForwardRef(TableCell)>
                  Temperature
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)>
                  Feels like
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)>
                  Atmospheric pressure
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)>
                  Humidity
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)>
                  Weather
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)>
                  Windspeed
                </ForwardRef(TableCell)>
              </ForwardRef(TableRow)>
            </ForwardRef(TableHead)>
            <ForwardRef(TableBody)>
              <ForwardRef(TableRow)
                key="2021-10-26 12:00:00"
              >
                <ForwardRef(TableCell)
                  component="th"
                  scope="row"
                >
                  12:00:00
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)>
                  12.37
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)>
                  11.93
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)>
                  1018
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)>
                  87
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)>
                  Clouds
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)>
                  2.09
                </ForwardRef(TableCell)>
              </ForwardRef(TableRow)>
              <ForwardRef(TableRow)
                key="2021-10-27 12:00:00"
              >
                <ForwardRef(TableCell)
                  component="th"
                  scope="row"
                >
                  12:00:00
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)>
                  12.38
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)>
                  11.98
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)>
                  1018
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)>
                  80
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)>
                  Rain
                </ForwardRef(TableCell)>
                <ForwardRef(TableCell)>
                  2.12
                </ForwardRef(TableCell)>
              </ForwardRef(TableRow)>
            </ForwardRef(TableBody)>
          </ForwardRef(Table)>
        </ForwardRef(Box)>
      </ForwardRef(Collapse)>
    </ForwardRef(TableCell)>
  </ForwardRef(TableRow)>
</Fragment>
`;
